lib/datasource/maven/index.ts

```diff
@@ -145,126 +124,139 @@ async function getSnapshotFullVersion(
+          const res = await isHttpResourceExists(artifactUrl);
+          const release: Release = { version };
+
+          if (is.string(res)) {
+            release.releaseTimestamp = res;
+          }
+
+          // Retry earlier for error status other than 404
+          if (res === null) {
+            retryEarlier = true;
+          }
+
+          workingReleaseMap[version] = res ? release : null;
+        }
+      );
+
+      await pAll(queue, { concurrency: 5 });
+      const cacheTTL = retryEarlier ? 60 : 24 * 60;
+      await packageCache.set(cacheNs, cacheKey, workingReleaseMap, cacheTTL);
+    }
+  }
 
-    // Retry earlier for status other than 404
-    const cacheTTL = Object.values(artifactsInfo).some((x) => x === null)
-      ? 60
-      : 24 * 60;
+  for (const version of Object.keys(releaseMap)) {
+    releaseMap[version] ||= workingReleaseMap[version] ?? null;
+  }
 
-    await packageCache.set(cacheNamespace, cacheKey, artifactsInfo, cacheTTL);
+  return releaseMap;
+}
+
+function getReleasesFromMap(releaseMap: ReleaseMap): Release[] {
+  const releases = Object.values(releaseMap).filter(Boolean);
+  if (releases.length) {
+    return releases;
   }
+  return Object.keys(releaseMap).map((version) => ({ version }));
+}
 
-  // Create releases for every version that exists in the repository
-  return versions
-    .filter((v) => artifactsInfo[v])
-    .map((version) => {
-      const release: Release = { version };
-      const releaseTimestamp = artifactsInfo[version];
-      if (releaseTimestamp && typeof releaseTimestamp === 'string') {
-        release.releaseTimestamp = releaseTimestamp;
-      }
-      return release;
-    });
+function isStableVersion(x: string): boolean {
+  return mavenVersion.isStable(x);
 }
 
-function getFallbackReleases(versions: string[]): Release[] | null {
-  const result = versions?.map((version) => ({
-    version,
-  }));
-  return result ?? null;
+function getLatestStableVersion(releases: Release[]): string | undefined {
+  return releases
+    .map(({ version }) => version)
+    .filter(isStableVersion)
+    .reduce((latestVersion, version) =>
+      compare(version, latestVersion) === 1 ? version : latestVersion
+    );
 }
 
 export async function getReleases({
   lookupName,
   registryUrl,
 }: GetReleasesConfig): Promise<ReleaseResult | null> {
-  const dependency = getDependencyParts(lookupName);
+  const mvnDep = getDependencyParts(lookupName);
   const repoUrl = registryUrl.replace(/\/?$/, '/'); // TODO #12070
 
-  logger.debug(`Looking up ${dependency.display} in repository ${repoUrl}`);
-
-  let releases: Release[] = null;
-  const versions = await getVersionsFromMetadata(dependency, repoUrl);
-  releases ||= await getReleasesFromHeadRequests(dependency, repoUrl, versions);
-  releases ||= getFallbackReleases(versions);
+  logger.debug(`Looking up ${mvnDep.display} in repository ${repoUrl}`);
 
+  let releaseMap = await fetchReleasesFromMetadata(mvnDep, repoUrl);
```

Comment:
Here we'll add HTML-based release info fetching, which will handle most of the Maven Central versions